kernel/combinatorics/hilb.cc

```diff
@@ -1400,5 +1426,733 @@ void hLookSeries(ideal S, intvec *modulweight, ideal Q, intvec *wdegree, ring ta
+	for(i=1;i<sz;i++)
+	{
+		s=MonoIdBasesSame(I,idorb[i]);
+    if(s)
+    {
+			ps=i+1;
+			break;
+		}
+	}
+
+	return(ps);
+}
+
+ideal MinimalMonomialBasis(ideal I)
+{
+	I=SortMonoId(I);
+	ideal J=idInit(1,1);
+	int i,k;
+	int count=0;
+	int sz=IDELEMS(I);
+	
+	std::vector<BOOLEAN> checks(sz,TRUE);
+  for(k=0; k<sz; k++)
+  {
+		if(checks[k])
+    {
+			idInsertOneMon(J,I->m[k]);
+      count++;
+      for(i=count;i<sz;i++)
+			{
+      	if(checks[i])
+        {
+        	if( pLmDivisibleByNoComp(J->m[count-1],I->m[i]))
+          {
+						checks[i]=FALSE;
+          }
+        }
+			}
+		}
+	}
+	return(J);
+} 
+    
+
+poly shiftInMon(poly p,int i, int lV, const ring r)
+{ 
+  /* shift the varibles in the  i^th layer*/
+	poly smon=p_One(r);
+  int j,sh,cnt;
+  cnt=r->N;
+  sh=i*lV;               
+ 
+  int *e=(int *)omAlloc((r->N+1)*sizeof(int));
+  int *s=(int *)omAlloc0((r->N+1)*sizeof(int));
+  p_GetExpV(p,e,r);
+
+  for(j=1;j<=cnt;j++)
+	{
+    if(e[j]==1)
+    {
+      s[j+sh]=e[j]; 
+    }
+  }
+	p_SetExpV(smon,s,r);
+  omFree(e);
+  omFree(s);
+  p_SetComp(smon, p_GetComp(p,r), r);
+	return(smon);
+}
+
+poly deleteInMon(poly w, int i, int lV, const ring r)
+{
+
+ /* delete the variables upto i^th block in 
+ * the monomial w*/
+	poly dw = p_One(currRing);
+	int *e = (int *)omAlloc((r->N+1)*sizeof(int));
+	int *s=(int *)omAlloc0((r->N+1)*sizeof(int));
+	p_GetExpV(w,e,r);
+	int j,cnt;
+	cnt=i*lV;
+ /*
+	for(j=1;j<=cnt;j++)
+	{	
+		e[j]=0; 
+  }*/
+  for(j=(cnt+1);j<(r->N+1);j++)
+    {
+        s[j]=e[j];
+    }
+  p_SetExpV(dw,s,r);//new exponents
+	omFree(e);
+	omFree(s);
+	p_SetComp(dw, p_GetComp(w,r), r);
+	return(dw);
+}
+
+void TwordMap(poly p, poly w, int lV, int d, ideal Jwi,bool &flag)
+{
+	int i;
+	//ring r=currRing;
+	poly smon,dw;
+	poly qmonp=NULL;
+
+	for(i=0;i<=d-1;i++)
+	{
+		dw=deleteInMon(w,i,lV,currRing);
+		smon=shiftInMon(p,i,lV,currRing);
+   
+		if(pLmDivisibleByNoComp(smon, w))
+		{
+			flag=TRUE;
+			return(idMakeOne(Jwi,p_One(currRing)));
+		}
+
+		if(pLmDivisibleByNoComp(dw, smon))
+		{
+			qmonp=p_Divide(smon,dw,currRing);
+			idInsertOneMon(Jwi,shiftInMon(qmonp,-d,lV,currRing));
+		}
+	}  
+	
+} 
+
+ideal colonIdeal(ideal S, poly w, int lV,ideal Jwi)
+{
+    if(idIs0(S))
+	{
+		return(S);
+	}
+	int i,j,d;
+  d=DegMon(w);  
+  bool flag=FALSE; 
+  int sz=IDELEMS(S);
+  int cnt=0;
+	for(i=0;i<sz;i++)
+  {
+		TwordMap(S->m[i],w,lV,d,Jwi,flag);
+		if(flag)
+		{
+			break;
+		}
+	}
+	Jwi = MinimalMonomialBasis(Jwi);
+	return(Jwi);
+}
+
+void HilbertSeries_OrbitData(ideal S, bool IG_CASE )
+{	
+	/* 
+	 * It is based on iterative right colon operation to the
+	 * monomial ideals of the free associative algebras.
+	 * The algorithm terminates for the monomial right
+	 * ideals whose monomials define regular formal language,
+	 * that is, all the monomials of ideal can be obtained from
+	 * finite subsets by applying the finite number
+	 * of elementary operations.  
+	 * */
+
+	int lV=currRing->N/(rBlocks(currRing)-3);
```

Comment:
really ? rBlocks refers to the internal representation of orderings and may be 1, 2, 3:
you need at least special requirement for the ordering